src/stdune/path.ml

```diff
@@ -1335,3 +1335,16 @@ let chmod ~mode ?(stats = None) ?(op = `Set) path =
         stats.st_perm land lnot mode
   in
   Unix.chmod (to_string path) mode
+
+let realpath t =
+  let path_s = to_absolute_filename t in
+  let stat = Unix.lstat path_s in
```

Comment:
Instead of checking the kind, we could catch Unix_error (EINVAL, _, _) and thus eliminate the unnecessary race condition.

Speaking of which, we should handle the other errors too (file or directory not existing etc.).

Reply:
> Instead of checking the kind, we could catch Unix_error (EINVAL, _, _) and thus eliminate the unnecessary race condition.

Good idea.

> Speaking of which, we should handle the other errors too (file or directory not existing etc.).

How should we handle them?